File: csharp/examples/HelloWorldPublisher/HelloWorldType.cs
using System;
using Int2Dds.Cdr;
using Int2Dds.Types;

namespace HelloWorldPublisher
{
    /// <summary>
    /// IDL struct: HelloWorld { unsigned long index; string message; };
    /// Extensibility: FINAL, non-keyed.
    /// </summary>
    [DdsType("HelloWorld", 0, false)]
    public class HelloWorld : IDdsType
    {
        public uint Index { get; set; }
        public string Message { get; set; } = "";

        public HelloWorld() { }

        public HelloWorld(uint index, string message)
        {
            Index = index;
            Message = message;
        }

        public byte[] SerializeCdr()
        {
            return SerializeCdr(false);
        }

        public byte[] SerializeCdr(bool xcdr2)
        {
            var w = new CdrWriter(Extensibility.Final);
            w.WriteU32(Index);
            w.WriteString(Message);
            return w.ToBytes();
        }

        public static HelloWorld DeserializeCdr(byte[] data)
        {
            var r = new CdrReader(data);
            return new HelloWorld(r.ReadU32(), r.ReadString());
        }

        public byte[] SerializeKey() => Array.Empty<byte>();
    }
}

File: python/examples/hello_world_type.py
"""
Example HelloWorld data type.

This is what int2dds-idl would generate from:

    struct HelloWorld {
        unsigned long index;
        string message;
    };
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar

from int2dds.cdr import CdrReader, CdrWriter, Extensibility
from int2dds.cdr.writer import CdrKeyWriter


@dataclass
class HelloWorld:
    """IDL struct: HelloWorld (Extensibility: FINAL)"""

    _dds_type_name: ClassVar[str] = "HelloWorld"
    _extensibility: ClassVar[Extensibility] = Extensibility.FINAL
    _has_key: ClassVar[bool] = False

    index: int = 0
    message: str = ""

    def _serialize_cdr(self) -> bytes:
        """Serialize to CDR bytes with encapsulation header."""
        w = CdrWriter(extensibility=self._extensibility)
        w.write_u32(self.index)
        w.write_string(self.message)
        return w.to_bytes()

    @classmethod
    def _deserialize_cdr(cls, data: bytes) -> "HelloWorld":
        """Deserialize from CDR bytes."""
        r = CdrReader(data)
        index = r.read_u32()
        message = r.read_string()
        return cls(index, message)

    def _serialize_key(self) -> bytes:
        """Serialize key fields only (big-endian, no encapsulation)."""
        return b""
